python/orca/tutorial/tensorflow/ncf_orca.py

```diff
@@ -0,0 +1,169 @@
+# data_type = "ml-25m"
+# Need spark3 to support delimiter with more than one character.
+start = time.time()
+spark= OrcaContext.get_spark_session()
+from pyspark.sql.types import StructField, StructType, StringType, IntegerType
+
+schema = StructType(
+    [
+        StructField('_c0', IntegerType(), True),
+        StructField('_c1', IntegerType(), True)
+    ]
+)
+df0 = spark.read.csv("{}/{}/ratings.csv".format(data_path, data_type), sep="::",schema=schema,
+                     header=False)
+df0 = df0.withColumnRenamed('_c0','user').withColumnRenamed('_c1','item')
+#print(int(df.describe("item").filter("summary = 'max'").select("item").first().asDict()['item']))
+min_user_id = df0.agg({"user": "min"}).collect()[0]["min(user)"]
+max_user_id = df0.agg({"user": "max"}).collect()[0]["max(user)"]
+min_item_id = df0.agg({"item": "min"}).collect()[0]["min(item)"]
+max_item_id = df0.agg({"item": "max"}).collect()[0]["max(item)"]
+print(min_user_id, max_user_id, min_item_id, max_item_id)
+from pyspark.sql import functions
+df = df0.withColumn('label', functions.lit(1))
+
+from pyspark.sql import functions as F
+from pyspark.sql import types as T
+import random
+
+def neg_sample(x):
+    item_count = len(x) * 4
+    max_count = max_item_id - len(set(x))
+    neg_item = random.sample(set(range(min_item_id, max_item_id+1)) - set(x), min(item_count,max_count))
+    return neg_item
+
+neg_sample_udf = F.udf(neg_sample, T.ArrayType(IntegerType(), False))
+
+df_neg= df.groupBy('user').agg(neg_sample_udf(F.collect_list('item')).alias('item_list'))
+from pyspark.sql.functions import *
+df_neg = df_neg.select(df_neg.user, explode(df_neg.item_list))
+df_neg = df_neg.withColumn('label', functions.lit(0))
+df = df.unionAll(df_neg)
+num_sample = df.count()
+train_df, test_df = df.randomSplit([0.8, 0.2],100)
+
+
+batch_size=1280
+epochs=2
+model_dir='./'
+
+# create an Estimator
+# backend = 'spark'
+# est = Estimator.from_keras(model_creator=model_creator, workers_per_node=1, backend=backend, model_dir=model_dir)
+est = Estimator.from_keras(model_creator=model_creator, workers_per_node=1)
+
+stats = est.fit(train_df,
+                epochs=epochs,
+                batch_size=batch_size,
+                feature_cols=['user', 'item'],
+                label_cols=['label'],
+                steps_per_epoch=int(800000 // batch_size),
+                validation_data=test_df,
+                validation_steps =int(200000 // batch_size))
+
+checkpoint_path = os.path.join(model_dir, "NCF.ckpt")
+est.save_checkpoint(checkpoint_path)
+
+
+# checkpoint_path = os.path.join(model_dir, "NCF.ckpt")
+# est.save_checkpoint(checkpoint_path)
+# print(model.get_weights())
+
+
+# evaluate with Estimator
+stats = est.evaluate(test_df, 
+                     feature_cols=['user', 'item'],
+                     label_cols=['label'],
+                     num_steps=int(0.2*num_sample // batch_size))
+
+print(stats)
+est.shutdown()
+end = time.time()
+print("Time used: ", end - start)
```

Comment:
May I ask what time information is counted here?

Reply:
I think we can just remove the timing in the tutorial?

Reply:
Sure.